gnc/sub8_perception/nodes/follow_orange_pipes.py

```diff
@@ -17,66 +17,162 @@ def __init__(self):
+        min_eigv = eig_vects[:, 1 - i] * np.sqrt(eig_vals[1 - i])
 
-        TODO:
-            Compute the kernel only once
-                (Too cheap to need)
-        '''
-        kernel = np.ones((scale, scale)) * -1
-        midpoint = (scale // 2, scale // 2)
-        cv2.circle(kernel, midpoint, midpoint[0], 1, -1)
+        # define center of pipe
+        center = np.average(wh, axis=1)
 
-        mean, std_dev = cv2.meanStdDev(image)
+        # define vertical vector (sub's current direction)
+        vert_vect = np.array([0.0, -1.0])
 
-        # Check if the scene is brighter than our a priori target
-        if mean > mean_thresh:
-            kernel = -kernel
+        if max_eigv[1] > 0:
+            max_eigv = -max_eigv
+            min_eigv = -min_eigv
 
-        normalized_cross_correlation = cv2.filter2D((image - mean) / std_dev, -1, kernel)
-        renormalized = normalized_cross_correlation
-        return renormalized
+        num = np.cross(max_eigv, vert_vect)
+        denom = np.linalg.norm(max_eigv) * np.linalg.norm(vert_vect)
+        angle_rad = np.arcsin(num / denom)
+
+        return center, angle_rad, [max_eigv, min_eigv]
+
+    def find_pipe_new(self, img):
+        #img[:, -100:] = 0
+        #img = cv2.GaussianBlur(img, (7, 7), 15)
+        last_image_timestamp = self.last_image_timestamp
+        hsv = cv2.cvtColor(img, cv2.COLOR_RGB2HSV)
+
+        lower = np.array([self.calculate_threshold(hsv), 0, 0])
+        upper = np.array([179, 255, 255])
+
+        # Take the threholded mask, remove noise, find the biggest contour, then draw a the best fit rectangle
+        #   around that box, finally use same algorithm on the best fit rectangle.
+        mask = cv2.inRange(hsv, lower, upper)
+        mask = cv2.morphologyEx(mask, cv2.MORPH_CLOSE, self.kernel)
+        contours, _ = cv2.findContours(np.copy(mask), cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
+        if len(contours) < 1:
+            print "None found"
+            return
+
+        # Find biggest area contour
+        self.last_draw_image = np.copy(hsv)
+        areas = [cv2.contourArea(c) for c in contours]
+        max_index = np.argmax(areas)
+        cnt = contours[max_index]
+
+        # Draw a miniAreaRect around the contour and find the area of that.
+        rect = cv2.minAreaRect(cnt)
+        box = cv2.cv.BoxPoints(rect)
+        box = np.int0(box)
+        mask = np.zeros(shape=mask.shape)
+        cv2.drawContours(mask, [box], 0, 255, -1)
+        rect_area = cv2.contourArea(box)
+
+        center, angle_rad, [max_eigv, min_eigv] = self.get_pose(mask)
+
+        # Check if the box is too big or small.
+        xy_position, height = self.occ_grid.get_tf(timestamp=last_image_timestamp)
+        expected_area = self.occ_grid.calculate_marker_area(height)
+        if expected_area * .3 < rect_area < expected_area * 2:
+            cv2.drawContours(self.last_draw_image, [box], 0, (255, 255, 255), -1)
+        else:
+            angle_rad = 0
```

Comment:
The idea here is that the marker is still in the frame just barley so we'll go to where we think it is, but the rotation is probably all wonky so just stay rotated like we are then fix once we get there.